File: proper_code/parking_lot/services/TicketBuilderService.ts
import { Ticket } from "../entities/Ticket";


class TicketBuilderService {

  private _ticketId?: string;
  private _spotId?: string;
  private _vehicleId?: string;
  private _parkedAt?: Date;
  private _entryGateId?: string;

  private static instance: TicketBuilderService;

  private constructor() {}

  static getInstance (): TicketBuilderService {
    return TicketBuilderService.instance ? TicketBuilderService.instance : TicketBuilderService.instance = new TicketBuilderService();
  }

  setTicketId(ticketId: string) {
    this._ticketId = ticketId;
    return this;
  }
  setSpotId(spotId: string) {
    this._spotId = spotId;
    return this;
  }
  setVehicleId(vehicleId: string) {
    this._vehicleId = vehicleId;
    return this;
  }
  setParkedAt(parkedAt: Date) {
    this._parkedAt = parkedAt;
    return this;
  }
  setEntryGate(entryGate: string) {
    this._entryGateId = entryGate;
    return this;
  }

  build(): Ticket {
    if (!this._entryGateId || !this._parkedAt || !this._spotId || !this._ticketId || !this._vehicleId) {
      throw new Error('Missing Fields to build Ticket')
    }
    return new Ticket(this._ticketId, this._spotId, this._entryGateId, this._parkedAt, this._vehicleId);
  }
}

export { TicketBuilderService };

File: proper_code/parking_lot_java/services/TicketBuilderService.java
package services;

import entities.Ticket;

import java.util.Date;

/**
 * Builder service for creating Ticket objects.
 * Note: This is NOT a singleton in the thread-safe version to avoid
 * shared mutable state between concurrent ticket creations.
 */
public class TicketBuilderService {
    private String ticketId;
    private String spotId;
    private String vehicleId;
    private Date parkedAt;
    private String entryGateId;

    public TicketBuilderService() {
    }

    public TicketBuilderService setTicketId(String ticketId) {
        this.ticketId = ticketId;
        return this;
    }

    public TicketBuilderService setSpotId(String spotId) {
        this.spotId = spotId;
        return this;
    }

    public TicketBuilderService setVehicleId(String vehicleId) {
        this.vehicleId = vehicleId;
        return this;
    }

    public TicketBuilderService setParkedAt(Date parkedAt) {
        this.parkedAt = parkedAt;
        return this;
    }

    public TicketBuilderService setEntryGateId(String entryGateId) {
        this.entryGateId = entryGateId;
        return this;
    }

    public Ticket build() {
        if (ticketId == null || spotId == null || vehicleId == null || parkedAt == null || entryGateId == null) {
            throw new IllegalStateException("Missing fields to build Ticket");
        }
        return new Ticket(ticketId, spotId, entryGateId, parkedAt, vehicleId);
    }
}
